### PR TITLE
feat: recipient address security module

### LINKED ISSUE #2014: [Tx risk mitigation] Unknown addresses/Safe on wrong chain module
## What is the feature about

A transaction risk mitigation (#1809) module to notify of interaction with "unused"/non-address book addresses.

### Unused address

- Not a smart contract
- No balance

### Non-address book address

- Not a smart contract
- Address not in address-book

### Safe on wrong chain

1. Check if address is smart contract contract on current chain.
2. If not, query whether address is Safe on all other chains in config.
3. If one of them is, show warning.

## The list of requirements

- [x] Create `SecurityModule`(s) in accordance with the interface and `SecurityResponse` type.
- [x] Add new module(s) to `SecurityModuleNames` and `SecurityModules`.
- [x] Create hook/context responsible for calling/storing the response of the module.

Depending on transaction modal redesign status:

- [ ] Implement `SecurityWarning`/`SecurityHint` components to render results.

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Resolves #2014

## How this PR fixes it

This adds the `RecipientAddressModule`, as well as associated hook and demo implementation in the transaction modal. It covers normal/multiSend transfers of native, ERC-20 and ERC-721 assets.

The implementation checks and warns of the following:

1. Whether the address is either an owner or present in the address book.
2. Whether the address is "used", e.g. it is an EoA with native balance.
3. Whether the address is a Safe deployed otherwise on mainnet (the user is sending funds to the wrong chain).

## How to test it

Create a transaction with the recipients matching either of the three above criteria and observe the warning(s).

## Screenshots

Please note that this is a demo implementation.

image

## Checklist
* [x] I've tested the branch on mobile 📱
* [x] I've documented how it affects the analytics (if at all) 📊
* [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/components/tx/TxSecurityWarnings/index.tsx [hunk 1]
```diff
@@ -15,10 +18,26 @@ export const TxSecurityWarnings = ({ safeTx }: { safeTx: SafeTransaction | undef
         redefineScanResult && (
           <>
             <BalanceChanges balanceChange={redefineScanResult.payload?.balanceChange} />
-            {redefineScanResult?.payload?.issues?.map((issue) => (
+            {redefineScanResult.payload?.issues?.map((issue) => (
               <SecurityHint key={issue.category} severity={issue.severity} text={issue.description.short} />
             ))}
-            <SecurityWarning severity={redefineScanResult?.severity} />
+            <SecurityWarning severity={redefineScanResult.severity} />
```
Comment (schmanu) [inline src/components/tx/TxSecurityWarnings/index.tsx:24]: We should sync after merging the PR or before on how to consolidate both results into one component.
They should appear in the same block.

**Review by schmanu — APPROVED** [review]

File: src/services/security/modules/RecipientAddressModule/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,1113 @@
...
+  ])
+}
+
+const getMockMultiSendCalldata = (recipients: Array<string>): string => {
+  // MultiSendCallOnly
+  const OPERATION = 0
+
+  const data = '0x'
+
+  const internalTransactions = recipients.map((recipient) => {
```
Comment (schmanu) [inline src/services/security/modules/RecipientAddressModule/index.test.ts:56]: The safe-core-sdk exports a util function  `encodeMultiSendData` for this.

File: src/services/security/modules/RecipientAddressModule/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,1113 @@
...
+              long: 'The address is not an owner or present in your address book and is not a smart contract',
+            },
+            type: 'UNKNOWN_ADDRESS',
+          },
+        ],
+      })
+    })
+  })
+
+  it('should not warn if the address(es) is/are used', async () => {
```
Comment (schmanu) [inline src/services/security/modules/RecipientAddressModule/index.test.ts:404]: What is the difference between this test and the previous one "should warn about recipient of native transfer recipients"?
They look identical to me.

File: src/components/tx/TxSecurityWarnings/index.tsx (hunk 1 above)
Comment (iamacook) [reply src/components/tx/TxSecurityWarnings/index.tsx:24]: I agree. `TxSecurityWarnings` is just for demonstrative purposes at the moment though.

File: src/services/security/modules/RecipientAddressModule/index.test.ts (hunk 2 above)
Comment (iamacook) [reply src/services/security/modules/RecipientAddressModule/index.test.ts:56]: I had/am having issues with using that helper. As this is purely testing related, I think it's alright keeping it as is. What do you think?

File: src/services/security/modules/RecipientAddressModule/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,1113 @@
...
+        data, // data
+      ],
+    )
+  })
+
+  const multiSendInterface = Multi_send__factory.createInterface()
+  return multiSendInterface.encodeFunctionData('multiSend', [ethers.utils.concat(internalTransactions)])
+}
+
+const createSafeTransaction = ({
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/services/security/modules/RecipientAddressModule/index.test.ts:73]: I've moved this and all related calldata mocking to `@/tests/transactions.ts`.

**Pull request merged**